bokehjs/src/lib/models/glyphs/webgl/webgl_utils.ts

```diff
@@ -0,0 +1,90 @@
+import {LineJoin} from "core/enums"
+import {Uniform} from "core/uniforms"
+import {color2rgba} from "core/util/color"
+import {uint32} from "core/types"
+import {HatchPattern} from "core/property_mixins"
+import {hatch_aliases} from "core/visuals/patterns"
+
+
+// WebGL shaders use integers for caps, joins and hatching.
+export const cap_lookup = {butt: 0, round: 1, square: 2}
+
+export const join_lookup = {miter: 0, round: 1, bevel: 2}
+
+const hatch_pattern_lookup: {[key: string]: number} = {
+  blank: 0,
+  dot: 1,
+  ring: 2,
+  horizontal_line: 3,
+  vertical_line: 4,
+  cross: 5,
+  horizontal_dash: 6,
+  vertical_dash: 7,
+  spiral: 8,
+  right_diagonal_line: 9,
+  left_diagonal_line: 10,
+  diagonal_cross: 11,
+  right_diagonal_dash: 12,
+  left_diagonal_dash: 13,
+  horizontal_wave: 14,
+  vertical_wave: 15,
+  criss_cross: 16,
+}
+
+
+function hatch_pattern_to_index(pattern: HatchPattern): number {
+  return hatch_pattern_lookup[hatch_aliases[pattern] ?? pattern] | 0
+}
+
+export function color_to_uint8_array(color_prop: Uniform<uint32>, alpha_prop: Uniform<number>): Uint8Array {
+  const ncolors: number = Math.max(color_prop.length, alpha_prop.length)
```

Comment:
It should be `Math.min()`, otherwise you may get out-of-bounds error, though at this point it would be a bug if those arrays had different lengths.

Reply:
Thanks for highlighting this.  My code here is fundamentally wrong, although it does render OK without any errors. The idea is that when rendering markers/quads, send the minimum amount of color data to the webgl shaders.  So if, say, the fill alpha and RGB color are the same for all instances (`UniformScalar`) then only send `ncolor=1` RGBAs to the shader, but if either vary (`UniformVector`) then send one each i.e. `ncolor=number_of_instances`. I had wrongly determined that if `color_prop` or `alpha_prop` were `UniformScalar` they would return a length of 1.  But they return the number of instances regardless, so the arguments to my `Math.max` (or your suggested `Math.min`) are always the same and equal to the number of instances, so this code always sends the maximum number of RGBAs to the shaders.

I've changed it to the correct logic which is
```javascript
const ncolors = color_prop.is_Scalar() && alpha_prop.is_Scalar() ? 1 : color_prop.length
```

Reply:
I implicitly assumed those were arrays and not uniforms, so obviously I was wrong..

Reply:
> I've changed it to the correct logic which is (...)

This is much better.